### PR TITLE
[DEV-1900] Api price-fee payment prediction

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
Migration

File: src/subdomains/core/buy-crypto/process/services/buy-crypto.service.ts [hunk 1]
```diff
@@ -220,8 +214,9 @@ export class BuyCryptoService {
   @Lock(7200)
   async process() {
     if (Config.processDisabled(Process.BUY_CRYPTO)) return;
-    if (!Config.processDisabled(Process.BUY_CRYPTO_AML_CHECK)) await this.doAmlCheck();
-    await this.buyCryptoBatchService.prepareTransactions();
+    if (!Config.processDisabled(Process.BUY_CRYPTO_AML_CHECK)) await this.buyCryptoPreparationService.doAmlCheck();
+    await this.buyCryptoPreparationService.refreshFeeAndPrice();
+    await this.buyCryptoPreparationService.prepareTransactions();
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/services/buy-crypto.service.ts:219]: Shouldn't the preparation be before fee/price refresh?

File: src/subdomains/core/sell-crypto/process/buy-fiat-preparation.service.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      .leftJoin('buyCrypto.cryptoRoute', 'cryptoRoute')
+      .leftJoin('cryptoRoute.user', 'user')
+      .where(`user.id = :userId`, { userId: In(userIds) })
+      .andWhere('cryptoInput.created = :date', { date: Between(dateFrom, dateTo) })
+      .andWhere('buyCrypto.amlCheck = :amlCheck', { amlCheck: CheckStatus.PASS })
+      .getRawOne<{ volume: number }>()
+      .then((result) => result.volume);
+
+    return { buy: buyVolume, convert: convertVolume };
+  }
```
Comment (davidleomay) [inline src/subdomains/core/sell-crypto/process/buy-fiat-preparation.service.ts:89]: Where is this private method used?

File: src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts [hunk 3]
```diff
@@ -382,6 +382,28 @@ export class BuyCrypto extends IEntity {
...
+    totalFeeAmount: number,
+    totalFeeAmountChf: number,
+  ): UpdateResult<BuyCrypto> {
+    const update: Partial<BuyCrypto> = {
+      absoluteFeeAmount: 0,
+      percentFee: fee,
+      percentFeeAmount: fee * this.inputReferenceAmount,
+      minFeeAmount,
+      minFeeAmountFiat: minFeeAmount,
+      totalFeeAmount,
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts:397]: `minFeeAmount` might be in crypto, if it's crypto-crypto, right?

File: src/subdomains/core/buy-crypto/process/services/buy-crypto-preparation.service.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+          minVolume,
+          userDataVolume.buy + userDataVolume.checkout, // + convert?
+          bankData?.userData,
+        ),
+      );
+    }
+  }
+
+  async refreshFeeAndPrice(): Promise<void> {
+    // Atm only for bankTx/checkoutTx BuyCrypto
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/services/buy-crypto-preparation.service.ts:85]: Rename to `setFee` (or similar), it doesn't set any prices.

File: src/subdomains/core/buy-crypto/process/services/buy-crypto-preparation.service.ts [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        'cryptoInput',
+        'cryptoRoute',
+        'cryptoRoute.user',
+        'cryptoRoute.user.userData',
+      ],
+    });
+
+    for (const entity of entities) {
+      // Skip convert buyCrypto
+      if (entity.cryptoInput) continue;
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/services/buy-crypto-preparation.service.ts:109]: Add filter to SQL query?

**Review by davidleomay — APPROVED** [review]

**Pull request merged**